### ISSUE TITLE
Variable validation is is completely ignored when a plan uses "-target" and type = object

### ISSUE THREAD

**Issue by cgetzen** [hashicorp/terraform#29600]
Tested on 1.0.7.

Smallest example:

```terraform
variable "my_var" {
  type = object({
    b = bool
  })

  default = {
    b = false
  }

  validation {
    condition = var.my_var.b == true
    error_message = "Wat."
  }
}
```
---
```
terraform plan

│ Error: Invalid value for variable
│
│   on .terraform/modules/default/variables.tf line 215:
│  215: variable "my_var" {
│
│ Wat.
│
│ This was checked by the validation rule at .terraform/modules/default/variables.tf:224,3-13.
```
---
```
terraform plan -target=module.default

# Completes plan, doesn't error out.
```

Planning without a target works.
Planning with a target on simple types (string, boolean) also works.

**Comment by alisdair** [comment]
Thanks for reporting this, @cgetzen. I can reproduce the issue on 1.0, but it appears to be fixed in the main branch and the [most recent 1.1.0 alpha build](https://releases.hashicorp.com/terraform/1.1.0-alpha20210908/). Can you try out that build and verify that the issue you're seeing will be resolved in 1.1.0?

**Comment by cgetzen** [comment]
Thanks for looking!

While 1.1.0 unblocks one use-case, it'd be nice to have different merge behavior. For example: 

```terraform
variable "ingress" {
  type = object({
    hooks = object({
      a = bool
      b = list(string)
    })
    more_settings = string
  })

  default = {
    hooks = {
      a = false
      b = []
    }
    more_settings = ""
  }

  validation {
    condition = var.ingress.hooks.a == false || length(var.ingress.hooks.b) > 0
    error_message = "A detailed explanation of why you can't have a without b."
  }
}
```

Instead, I get this error message:
```
 The given value is not suitable for child module variable "ingress" defined at .terraform/modules/default/variables.tf:148,1-19: attribute "hooks": attribute "b" is required.
```

So, validation doesn't really work here because the default `b = []` disappeared.
That's not too concerning because in this case, it's doing what I want, albeit with a difficult to read error message.

I could see other use-cases where the default is, for example `b = ["some actual default values"]`, and 1.1.0 will break this use-case.

I guess the question is: When you have object variables with defaults, does specifying the variable in a module overwrite all the defaults, even if some weren't explicitly set in the module?
I would love to see the input variables be merged on top of the defaults.

**Comment by apparentlymart** [comment]
Hi @cgetzen,

It seems like you've confirmed that your original bug report is fixed ready to be included in Terraform v1.1, so I'm going to close this issue.

I'd be happy to discuss your follow-on questions about the interactions between type constraints and default values if you start a topic about it in [the community forum](https://discuss.hashicorp.com/c/terraform-core/27). However, the short answer is that a `default` value applies only if the caller doesn't set a value for the variable at all, and so indeed if they provide a new value that doesn't match the type constraint and validation rules then they'll see an error.
